### ISSUE TITLE
Long messages are not truncated upon display

### ISSUE THREAD

**Issue by joker314** [decent-chat/decent#280]
## Expected Behaviour

When you receive a really long message, it shouldn't take up the whole screen, because that's annoying

## Actual behaviour

The full message is displayed

## Steps to reproduce

1) Go to a server where you can write messages
2) Send a message which is very very long
3) Observe that it takes up all the space

## Suggested fix

This needs discussion. We want to be able to allow paragraphed and mature discussion; but we don't want people to be able to spam. Options from other sites:

- [Discord] Character limit + Message rate limiting (not effective, but it works for them)
- [YouTube] "Read More" button when height of message exceeds certain value (very effective - as it allows mature discussion; drawback is that once you open a message you have to scroll to the bottom to close it)
- [My own] Use Youtube's method, but have "Read more" open a sort-of pop-up which can be closed with an "X". The pop-up has a scrollbar, but the "X" is always in view
- [My own] Mix Youtube's and Discord's methods. Have messages be restricted in length, yet still shorten them when displayed unless the user wants to do something else. This also may prevent people taking up the entire server memory, etc.

[[ Needs discussion ]]

**Comment by towerofnix** [comment]
Ping @PullJosh especially for design input.

**Comment by towerofnix** [comment]
Taking up server memory is unfortunately not something we've dealt with at all, yet... it's very easy to quickly send tons of long, space-taking messages. (Or, more directly, to upload many images, but that's another topic.)

Would a combination of your third and fourth suggestions work? I love the third idea for its UI but I wonder if it makes sense to have a (very long) character limit as well.

**Comment by heyitsmeuralex** [comment]
We can make the character limit a server setting?

**Comment by heyitsmeuralex** [comment]
We should also truncate message length in the _pins_ and _mentions_ tabs in the right sidebar. Because there's so little space for the message even not-too-long ones end up taking loads of vertical space.

**Comment by joker314** [comment]
I like the server setting idea, but as Flo said, we'll need to have a hard think about how to ensure people can't crash the server.

I agree that we should definitely cut the messages off in pins and mentions, but we should be careful about languages where different characters are used depending on where a letter appears in a word.

**Comment by heyitsmeuralex** [comment]
>we'll need to have a hard think about how to ensure people can't crash the server

If a server admin sets the limit to `9000000`, it's not our problem. I want to allow server owners to have as much freedom as possible.

**Comment by joker314** [comment]
Okay, but what about writing several messages successively, does Decent have adequate rate limiting?

**Comment by heyitsmeuralex** [comment]
Decent has no rate-limiting as of yet.

**Comment by joker314** [comment]
Right, this should probably change at some point. Has a separate issue been filed?

**Comment by towerofnix** [comment]
Rate limiting is almost a separate topic (though related). I don't think there's an issue for it yet. One thought on how it should be implemented is how to prevent users from simply signing up another account and spamming with that one (if we limit per-account). Except for on authorized servers / servers where users have no write permission by default, anybody can post data simply by registering.

**Comment by heyitsmeuralex** [comment]
We could rate-limit the register endpoint at 1 successful request/day or something?

**Comment by joker314** [comment]
Per IP? People can use proxy scripts and such, it's not too likely to work. You could try and place a CAPTCHA restriction?

**Comment by heyitsmeuralex** [comment]
@towerofnix note that there _are_ other CAPTCHA providers than Google, although theirs is objectively the best. Regardless, I want registers to be open so bots can register themselves on a server.
